File: java21/src/main/java/io/github/sps23/designpatterns/prototype/Document.java
package io.github.sps23.designpatterns.prototype;

import java.util.ArrayList;
import java.util.List;

/**
 * Prototype pattern in Java 21.
 *
 * <p>
 * Demonstrates both a shallow copy via the classic {@link Cloneable} mechanism
 * and a safe deep copy via a dedicated method. The {@code sections} list is the
 * mutable nested state that separates the two copying strategies.
 * </p>
 */
public class Document implements Cloneable {

    private String title;
    private String author;
    private List<String> sections;

    public Document(String title, String author, List<String> sections) {
        this.title = title;
        this.author = author;
        this.sections = new ArrayList<>(sections);
    }

    /**
     * Shallow copy. The new {@code Document} shares the same {@code sections} list
     * reference with the original, which is the classic Java prototype behaviour
     * when relying on {@link Object#clone()}.
     */
    @Override
    public Document clone() {
        try {
            return (Document) super.clone();
        } catch (CloneNotSupportedException e) {
            // This should never happen because we implement Cloneable.
            throw new AssertionError(e);
        }
    }

    /**
     * Deep copy. Creates a new list so mutations on the copy do not leak back to
     * the original document.
     */
    public Document deepCopy() {
        return new Document(title, author, new ArrayList<>(sections));
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getAuthor() {
        return author;
    }

    public void setAuthor(String author) {
        this.author = author;
    }

    public List<String> getSections() {
        return sections;
    }

    public void setSections(List<String> sections) {
        this.sections = sections;
    }
}

File: kotlin/src/main/kotlin/io/github/sps23/designpatterns/prototype/Document.kt
package io.github.sps23.designpatterns.prototype

/**
 * Prototype pattern in Kotlin.
 *
 * Data classes provide a generated `copy()` method, which is Kotlin's idiomatic
 * prototype mechanism. Because the generated copy keeps the same mutable list
 * reference by default, the example also shows a `deepCopy()` helper for cases
 * where independent nested state is required.
 */
data class Document(
    val title: String,
    val author: String,
    val sections: MutableList<String> = mutableListOf(),
) {
    /** Creates a copy with an independent sections list. */
    fun deepCopy(): Document = copy(sections = sections.toMutableList())
}
